internal/charmstore/migrations_test.go

```diff
@@ -822,49 +1471,104 @@ func (s *migrationsSuite) checkBaseEntitiesCount(c *gc.C, expectCount int) {
 	c.Assert(count, gc.Equals, expectCount)
 }
 
-func (s *migrationsSuite) insertEntity(c *gc.C, id *charm.Reference, name string, size int64) {
-	entity := &mongodoc.Entity{
-		URL:     id,
-		BaseURL: baseURL(id),
-		Name:    name,
-		Size:    size,
+// insertEntity creates a new entity in the database based on the
+// provided mongodoc.Entity. As a convenience some fields will
+// automatically derived if they are set to their zero value, these are:
+//
+//     BaseURL
+//     Name
+//     Revision
+//     Series
+//     User
+//     PromulgatedRevision
+//
+// After insertion any name in fields will be removed from the inserted
+// document.
+func (s *migrationsSuite) insertEntity(c *gc.C, entity *mongodoc.Entity, fields ...string) {
+	if entity.BaseURL == nil {
+		entity.BaseURL = baseURL(entity.URL)
+	}
+	if entity.Name == "" {
+		entity.Name = entity.URL.Name
+	}
+	if entity.Revision == 0 {
+		entity.Revision = entity.URL.Revision
+	}
+	if entity.Series == "" {
+		entity.Series = entity.URL.Series
+	}
+	if entity.User == "" {
+		entity.User = entity.URL.User
+	}
+	if entity.PromulgatedRevision == 0 {
+		if entity.PromulgatedURL != nil {
+			entity.PromulgatedRevision = entity.PromulgatedURL.Revision
+		} else {
+			entity.PromulgatedRevision = -1
+		}
 	}
 	err := s.db.Entities().Insert(entity)
 	c.Assert(err, gc.IsNil)
 
-	// Remove the denormalized fields if required.
-	if name != "" {
+	// Remove required fields
+	var remove bson.D
+	for _, f := range fields {
+		remove = append(remove, bson.DocElem{f, true})
+	}
+	if len(remove) == 0 {
 		return
 	}
-	err = s.db.Entities().UpdateId(id, bson.D{{
-		"$unset", bson.D{
-			{"user", true},
-			{"name", true},
-			{"revision", true},
-			{"series", true},
-		},
-	}})
+	err = s.db.Entities().UpdateId(entity.URL, bson.D{{"$unset", remove}})
 	c.Assert(err, gc.IsNil)
 }
 
-func (s *migrationsSuite) insertBaseEntity(c *gc.C, id *charm.Reference, acls *mongodoc.ACL) {
-	entity := &mongodoc.BaseEntity{
-		URL:    id,
-		Name:   id.Name,
-		User:   id.User,
-		Public: true,
+// Standard entity versions. These are lists of fields that can be used with
+// insertEntity to create database documents from previous deployments.
+var (
```

Comment:
Nice, Martin. +1 on both, this being really cool solution and adding "fields" postfix.